added backup

19ngaddam opened the pull request.

19ngaddam committed:
added backup

dmpotter44 commented on src/trialbackup.ts:
```diff
@@ -0,0 +1,130 @@
+    try {
+      const request = https.get(url, function (response) {
+        response.pipe(file).on('close', () => {
+          exec('unzip src/backup -d src/backups/', (error, stdout, stderr) => {
+            if (error) console.log(error);
+            resolve();
+          });
```
This should probably unzip using a Node library and not an external ZIP executable. It sounds like [yauzl](https://github.com/thejoshwolfe/yauzl) is the most popular Node module.

dmpotter44 commented on src/trialbackup.ts:
```diff
@@ -0,0 +1,130 @@
+  });
+}
+
+export function getBackupCriteria(trial: TrialBackup): string {
+  const criteria: string = trial.clinical_study.eligibility.criteria.textblock;
+  return criteria;
+}
+
+export function getBackupSummary(trial: TrialBackup): string {
+  const summary: string = trial.clinical_study.brief_summary.textblock;
+  return summary;
+}
+
+export function getBackupPhase(trial: TrialBackup): string {
+  const phase: string = trial.clinical_study.phase;
+  return phase;
+}
+
+export function getBackupStudyType(trial: TrialBackup): string {
+  const studytype: string = trial.clinical_study.study_type;
+  return studytype;
+}
```
Rather than write functions that provide simple accessors that someone could presumably handle themselves, it seems like it would make sense to provide a function that takes a ResearchStudy and adds any missing data itself.

dmpotter44 commented on src/trialbackup.ts:
```diff
@@ -0,0 +1,130 @@
+}
+
+export function getBackupTrial(nctId: string): TrialBackup {
+  const filePath = `src/AllPublicXML/${nctId.substr(0, 7)}xxxx/${nctId}.xml`;
```
We can't assume `src/AllPublicXML` exists here. As this is intended to be a library, we have no control over the folder structure of the caller. It would likely make more sense to require the caller to provide a directory to use to store data.

dmpotter44 commented on src/trialbackup.ts:
```diff
@@ -0,0 +1,130 @@
+  return json;
+}
+export function getDownloadedTrial(nctId: string): TrialBackup {
+  const filePath = `src/backups/${nctId}.xml`;
```
Same as the previous comment: it would make more sense to require the caller to provide the directory to save data to.

dmpotter44 commented on src/trialbackup.ts:
```diff
@@ -0,0 +1,130 @@
+}
+export function downloadRemoteBackups(ids: string[]) {
+  let url = 'https://clinicaltrials.gov/ct2/download_studies?term=';
+  for (const id of ids) {
+    url += `${id}+OR+`;
+  }
+  //remove trailing +OR+
+  url = url.slice(0, -4);
```
Or just:
```javascript
const url = 'https://clinicaltrials.gov/ct2/download_studies?term=' + ids.join('+OR+');
```
However, we should probably also be ensuring the IDs are properly escaped for a URL and a search query. We could just `map(encodeURIComponent)` but it may make more sense to check and confirm that each individual ID won't mess up the query.

dmpotter44 commented on src/trialbackup.ts:
```diff
@@ -0,0 +1,130 @@
+  //remove trailing +OR+
+  url = url.slice(0, -4);
+  console.log(url);
+  const file = fs.createWriteStream('src/backup.zip');
```
For this we should probably use a temp file. `os.tmpdir()` gets the official temporary directory, but we may want to look into using a library for generating temporary files so that multiple instances can be running at once.

19ngaddam committed:
Added testing suite for code coverage on the backup code

dmpotter44 commented:
I think it would likely make sense to refactor this into an object that takes on construction a scratch directory where it can store the downloaded clinical trials, and then can be used to update `ResearchStudy` objects with missing data. So an API that's something like:

```typescript
/**
 * This is the service that handles fetching data.
 */
class ClinicalTrialGov {
  constructor(public dataDir: string) {}
  retrieveClinicalTrial(ids: string[]): Promise<TrialBackup>;
}

/**
 * Class that implements the existing interface but adds a helper method
 */
class TrialBackup {
  addMissingData(study: ResearchStudy): void;
}
```

Note that implementations are intentionally missing, so this isn't valid code.

In the future we might want to add more functionality to the base service to deal with things like cleaning up the data directory and caching trials so that a second request for existing trials doesn't generate a new web request.

dmpotter44 committed:
Refactor the FHIR types

It was always intended that the FHIR types would be replaced by an
external library, this helps work towards that goal and also resolves
instances where the ResearchStudy FHIR type is needed but not the
ResearchStudy helper class.

dmpotter44 committed:
Fix (sort of) tests

19ngaddam committed:
added code structure changes

19ngaddam committed:
updated testing suite accordingly

19ngaddam committed:
Use passed in file path. unzip module still unrecognized

dmpotter44 committed:
Fix dependencies/package import/formatting

19ngaddam committed:
another unzipper not recognized by typescript!

19ngaddam committed:
got extract-zip to work

19ngaddam committed:
linting fix

19ngaddam committed:
added more test cases for code coverage

19ngaddam committed:
remove more fields from sample research study to test more code coverage

dmpotter44 committed:
Remove unnecessary dependency

dmpotter44 committed:
Merge branch 'backup' of github.com:mcode/clinical-trial-matching-service into backup

dmpotter44 committed:
Update config/fix code formatting

dmpotter44 commented on spec/backup.spec.ts:
```diff
@@ -0,0 +1,147 @@
+  let trial : trialbackup.TrialBackup;
+  beforeAll(async function () {
+    
+    const downloader = new trialbackup.ClinicalTrialGov(filepath)
```
We should figure out how to mock the download so that we can download a known-good version and aren't relying on a network connection and ClinicalTrial.gov being up.

19ngaddam committed:
Updated code so searchset uses research study interface instead of class

19ngaddam committed:
added more test cases fro code coverage

19ngaddam committed:
added test case for more code cov

dmpotter44 committed:
Refactoring and renaming, also actually export

dmpotter44 committed:
More refactoring; test coverage to 100%

Specifically, this splits a few functions to make mocking and testing
various parts of the backup service easier.

dmpotter44 committed:
Re-enable cleaning up temp files
